File: study1/solutions/week5/binary_search.py
# https://leetcode.com/problems/binary-search/description/


# import bisect
# class Solution:
#     def search(self, nums: List[int], target: int) -> int:
#         index = bisect.bisect_right(nums, target)
#         return index - 1 if index > 0 and nums[index - 1] == target else -1

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        start = 0
        end = len(nums) - 1
        current = int((start + end)/2)
        while nums[current] != target:
            if start == end:
                return -1

            if target < nums[current]:
                end = min(current, end - 1)
                current = int((start + end)/2)
                continue
            
            start = max(current, start + 1)
            current = int((start + end)/2)

        return current

File: study1/week5/binary-search.java
// https://leetcode.com/problems/binary-search/
// Time Complexity: O(logn)
// Space Complexity: O(1)
class Solution {
    public int search(int[] nums, int target) {
        int left = 0;
        int right = nums.length-1;

        while (left <= right) {
            int mid = (left + right)/2;
            System.out.println("mid: " + nums[mid] + ", left: " + nums[left] + ", right: " + nums[right]);
            if (nums[mid] == target) {
                return mid;
            } else if (nums[mid] < target) {
                left = mid+1;
            } else {
                right = mid-1;
            }
        }

        return -1;
    }
}